### PR TITLE
Improve python interface for the solver

### PR DESCRIPTION
**Pull request by nitnelave**
With these additions, it becomes possible to use the solver to handle the learning rate while setting the inputs/diffs in python.

### CODE REVIEW

File: python/caffe/test/test_solver.py [hunk 1]
```diff
@@ -38,6 +38,25 @@ def test_solve(self):
...
+        data[...] = 0
+        # Set the diffs to 1
+        net.layers[1].blobs[0].diff[...] = 1
+        # Apply the update, the initial learning rate should be 0.01
+        self.solver.apply_update()
+        # Check that the new weights are -0.01, with a precision of 1e-7
+        self.assertTrue((data - -0.01 * np.ones(data.shape)).max() < 1e-7)
+
+
+
```
Comment (longjon) [inline python/caffe/test/test_solver.py]: Why so much space?

File: python/caffe/test/test_solver.py [hunk 2]
```diff
@@ -38,6 +38,25 @@ def test_solve(self):
...
+        self.assertEqual(self.solver.iter, 0)
+        self.solver.set_iter(50)
+        self.assertEqual(self.solver.iter, 50)
+
+    def test_apply_update(self):
+        net = self.solver.net
+        data = net.layers[1].blobs[0].data[...]
+        # Reset the weights of that layer to 0
+        data[...] = 0
+        # Set the diffs to 1
```
Comment (longjon) [inline python/caffe/test/test_solver.py]: No need to write comments that don't say anything more than the code.

**Comment by longjon** [comment]
Seems like a bit of an odd case, but I guess it makes sense; one might also want to change the iter for debugging or other purposes.

**Comment by nitnelave** [comment]
@longjon updated as required

File: include/caffe/solver.hpp [hunk 3]
```diff
@@ -73,6 +73,7 @@ class Solver {
     return test_nets_;
   }
   int iter() { return iter_; }
+  void SetIter(int iter) { iter_ = iter; }
```
Comment (longjon) [inline include/caffe/solver.hpp]: Ah, I noticed this before but neglected to comment (sorry!): this should be `set_iter`; trivial functions use underscore_case (like the getter above).

**Comment by nitnelave** [comment]
@longjon fixed function name.

**Comment by Noiredd** [comment]
This has also been attempted in #6238, slightly more clever with the `iter++` IMO, but has no tests (for now).

**Comment by shelhamer** [comment]
Thanks for this @nitnelave, but a fuller interface has been exposed with the merge of #6238. Note that it includes c5b5b55845c20930026f1f563136247c41f58075 for the update test, with your authorship of course.

**Pull request closed without merging**